fix: Fixes for a few things discovered from the message ordering refactor:

ShadauxCat opened the pull request:
- A try/catch block in ProcessReceiveQueue was accidentally removed and has been added back.
- Some buffers were being copied up to buffer.Length instead of buffer.Position, so they were copying too much data.
- PooledNetworkBuffer's implementation of Dispose() was incorrect, and caused anything that had a compile-time static type of Stream or NetworkBuffer to call the parent class implementation, resulting in bad behavior and leaked buffers.

ShadauxCat committed:
Fixes for a few things discovered from the message ordering refactor:

- A try/catch block in ProcessReceiveQueue was accidentally removed and has been added back.
- Some buffers were being copied up to buffer.Length instead of buffer.Position, so they were copying too much data.
- PooledNetworkBuffer's implementation of Dispose() was incorrect, and caused anything that had a compile-time static type of Stream or NetworkBuffer to call the parent class implementation, resulting in bad behavior and leaked buffers.

0xFA11 commented on com.unity.multiplayer.mlapi/Runtime/Serialization/Pooled/PooledNetworkBuffer.cs:
```diff
@@ -24,8 +24,17 @@ public static PooledNetworkBuffer Get()
         /// Returns the PooledNetworkBuffer into the static NetworkBufferPool
+        /// Called by Dispose in the parent class implementation.
+        /// We cannot override Dispose because it's not declared virtual
+        /// And if we override it via `public new void Dispose()` then it doesn't get called
+        /// on anything with a static type other than PooledNetworkBuffer, which then results in a leak:
+        ///
+        /// Stream buffer = PooledNetworkBuffer.Get();
+        /// buffer.Dispose();
+        /// 
+        /// ^ Static type is Stream, this calls Stream::Dispose() instead of PooledNetworkBuffer::Dispose()
         /// </summary>
-        public new void Dispose()
+        public override void Close()
```
I like this more!

mattwalsh-unity commented on com.unity.multiplayer.mlapi/Runtime/Messaging/MessageQueue/MessageQueueProcessor.cs:
```diff
@@ -216,7 +216,19 @@ public void ProcessReceiveQueue(NetworkUpdateStage currentStage, bool isTesting)
+                            {
+                                ProcessMessage(currentQueueItem);
+                            }
+                            catch (Exception ex)
```
Ok, but do we know what specific exception could happen here, and if so can we catch that one?

And actually, when I look at the `ProcessMessage` impl, it already has a try/catch around the entire body (and also for top-most "Exception").  How could we ever get into the catch block?

ShadauxCat replied:
To answer the second question first: Yes, I realized that as well, and that's why I'm going to back this out. :)

To answer the first question second: `ProcessMessage` can call `InvokeRpc` which goes into user code which can throw literally anything at any time. This try/catch block is to make sure exceptions thrown from user code don't result in things like leaked PooledNetworkBuffers and don't result in other messages in the queue being dropped on the floor because an exception from one RPC halted the processing of the queue.

NoelStephensUnity replied:
And doesn't break the current stack which would end up never call this:
```
                    //We call this to dispose of the shared stream writer and stream
                    currentFrame.CloseQueue();
```
Which will never properly close the queue and also drop all of the remaining messages to be processed.

mattwalsh-unity commented on com.unity.multiplayer.mlapi/Runtime/Serialization/Pooled/PooledNetworkBuffer.cs:
```diff
@@ -24,8 +24,17 @@ public static PooledNetworkBuffer Get()
         /// <summary>
         /// Returns the PooledNetworkBuffer into the static NetworkBufferPool
+        /// Called by Dispose in the parent class implementation.
+        /// We cannot override Dispose because it's not declared virtual
```
Thank you for this anticipatory comment

ShadauxCat committed:
A couple more fixes:
- Process despawns always at PostLateUpdate
- Don't use fragmented channels for messages that don't actually need to fragment - UNet behaves poorly under load with certain use cases, which message batching has unfortunately invoked.

ShadauxCat committed:
Fixed tests.

ShadauxCat committed:
Switched a magic number to a constant and added a comment.

NoelStephensUnity reviewed (approved):
Looks like some good changes to me!

jeffreyrainy commented on com.unity.multiplayer.mlapi/Runtime/Messaging/MessageQueue/MessageQueueProcessor.cs:
```diff
@@ -314,8 +317,17 @@ private void SendCallback(ulong clientId, MessageBatcher.SendStream sendStream)
+            // if we have no other choice.
+            if (length > k_FragmentationThreshold)
+            {
+                channel = NetworkChannel.Fragmented;
```
I don't think you can do that. As far as I'm aware there's no ordering guarantees **between** channels. As such, stuff that was already sent on the non-fragmented channel could arrive after later fragmented message on the fragmented channel. If we get there and have this issue, I'd say let's switch it all to fragmented, and drop the condition.

ShadauxCat replied:
Unfortunately, the message that hits this is the very first message we send, so... that'd just result in everything being fragmented.

LukeStampfli replied:
So what does this change do then? We still end up sending everything on the reliableFragementedSequenced channel which is what was happening already before?

ShadauxCat replied:
ConnectionAccepted gets switched to Fragmented. I haven't seen any other messages that get too large for ReliableSequenced so far, though it's possible there are some. But this changelist basically changes everything to ReliableSequenced... and then this if check here is just to handle ConnectionAccepted and nothing else.

In my serialization RFC I'm going to rework that and any other messages that can get that large to be multiple messages instead of one mega-message, so this should stop being an issue for internal messages and will only matter for user RPCs or CustomMessages that get too large.

ShadauxCat replied:
So my response here is to say:

- This is a short-term fix to address the issues we're seeing in develop now, and will be torn out again soon to be fixed a better way with the serialization refactor
- For internal messages, we won't worry about ordering because we'll make it a requirement going forward to keep them within MTU size and not use fragmented messages
- For user-generated messages (RPCs) we actually currently don't accept anything larger than MTU, but in the serialization refactor we'll start accepting them. We'll make sure it's documented that large messages that need to be fragmented may be sequenced separately from small ones that don't need to be.

jeffreyrainy replied:
I don't see why it's not possible to simply use `channel = NetworkChannel.Fragmented` always. What's the reason against ?

ShadauxCat replied:
See my comment above in reply to Luke. Fragmented is not able to handle high volumes. It starts dropping messages.

jeffreyrainy replied:
yikes. Please, please, let's not forget this. But yeah, as a temporary band-aid...

LukeStampfli commented on com.unity.multiplayer.mlapi/Runtime/Messaging/MessageQueue/MessageQueueProcessor.cs:
```diff
@@ -325,12 +337,21 @@ private void SendCallback(ulong clientId, MessageBatcher.SendStream sendStream)
         {
+            var channel = item.NetworkChannel;
+            // If the length is greater than the fragmented threshold, switch to a fragmented channel.
+            // This is kind of a hack to get around issues with certain usages patterns on fragmentation with UNet.
```
I'm curious is there a Slack thread or something to learn more about that UNet fragmentation issue?

ShadauxCat replied:
No slack thread. But I'll tell you the behavior we see...

With UNet delivering things using ReliableFragmentedSequenced and using batched messages at around 530 bytes per batch, UNet runs out of buffer space in SceneTransitioningTest with as little as 6 objects spawning per second and starts dropping packets.

If we turn off batching - sending the same amount of overall data but in smaller 40 byte chunks instead of combined 530 byte chunks - it can get up to ~80 per second before packtes start getting dropped.

If we send on ReliableSequenced, it got up to 550 per second (the maximum the slider would allow) and nothing got dropped.

So 1) Fragmented is dropping messages that the network can handle without issue, and 2) for some reason sending the same data in fewer, larger chunks causes it to drop them more aggressively than sending in more, smaller chunks, even with the same amount of overall bandwidth usage.

I don't know where to find the UNet source to try to better understand why that happens, but I think we should avoid using ReliableFragmentedSequenced as much as possible.

LukeStampfli replied:
I guess that kind of makes sense. Maybe you could raise this in the internal unet slack channel just so that the right people know that this is an existing issue. And maybe they even have a fix for it. Ignore this if you already did.

ShadauxCat replied:
What is the internal unet channel? I will go raise it there.

ShadauxCat committed:
Merge branch 'develop' into fix/message_ordering_quick_fixes
